src/main/scala/GuessWho/GuessWho.scala

```diff
@@ -15,29 +17,60 @@ object GuessWho extends App{
 
   private val characterList:Seq[Character] = Seq(jon, tony, bill, paul, gary, steve, jess, vicky, sarah, beff)
 
+
   private val game:GameBoard = new GameBoard(characters = characterList, defaultChosenCharacter = None)
+  private val interface: Interface = new Interface()
+
 
   private var _playersRemaining:Seq[Character] = game.getRemainingCharacters
+  private var _newTurn:Boolean = true
+  private var _winner:Boolean = false
+
+  def print_characters() = if(_newTurn){
+    interface.displayCharacters(_playersRemaining)
+    _newTurn = false
+  }
+
+  def filter_by_name(): Seq[Character] = {
+    val name: String = interface.get_name_choice()
+    val filteredPlayers = game.filterRemaining(12, name)
+    _playersRemaining = _playersRemaining.filter(x => filteredPlayers.exists(_.name == x.name))
+    _winner = if (_playersRemaining.length == 1) true else false
+    _playersRemaining
+  }
 
 
-  def displayCharacters(characters:Seq[Character]):Unit = {
-    _playersRemaining.foreach { character =>
-      println(
-        s"""
-           |Name: ${character.name}
-           |Has Hair: ${character.hasHair}
-           |Has Facial Hair: ${character.hasFacialHair}
-           |Has Glasses: ${character.hasGlasses}
-           |Has Hat: ${character.hasHat}
-           |Gender: ${character.gender}
-           |Eye Colour: ${character.eyeColour}
-           |Hair Colour: ${character.hairColour}
-    """.stripMargin
-      )
+  def filter_by_attribute(attribute:Int):Seq[Character] = {
+    val guessNumber:Int = interface.check_user_guess(attribute)
+    guessNumber match {
+      case 0 =>{
+        println("Back to attribute options")
+        user_turn()
+      }
+      case -1 => {
+        println("Invalid option, please choose a valid option e.g. '1' ")
+        user_turn()
+      }
+      case _ => guessNumber
     }
-    println(s"Number of remaining characters: ${_playersRemaining.length}")
+    val filteredPlayers = game.filterRemaining(guessNumber)
+    _playersRemaining = _playersRemaining.filter(x => filteredPlayers.exists(_.name == x.name))
+    _playersRemaining
   }
 
-  displayCharacters(_playersRemaining)
+  def user_turn():Boolean = {
```

Comment:
Since this is recursion, keep all the variables you're using for it inside of it!

```scala
def user_turn(playersRemaining: Seq[Character], chosenCharacter: Character): Boolean = {
    print_characters()
    val attribute = interface.get_attribute_choice()

    attribute match {
      case 1 =>
        val guessedCharacter: Character = ???
        if (guessedCharacter == chosenCharacter) {
          println("You win")
          true
        } else {
          user_turn(playersRemaining = filter_by_name(playersRemaining), chosenCharacter = ???)
        }
      case other => user_turn(playersRemaining = filter_by_attribute(playersRemaining, other), chosenCharacter = ???)
    }
```

My attempt is by no means perfect, but it should show you how keeping the values inside a recursive loop makes this a lot easier to follow!